src/windows.rs

```diff
@@ -0,0 +1,343 @@
+/// functions by POSIX). These specifically do *not* contain mutexes and memory
+/// allocation/deallocation. They *do* contain routines to terminate the program, to further
+/// manipulate signals (by the low-level functions, not by this library) and to read and write file
+/// descriptors. Calling program's own functions consisting only of these is OK, as is manipulating
+/// program's variables ‒ however, as the action can be called on any thread that does not have the
+/// given signal masked (by default no signal is masked on any thread), and mutexes are a no-go,
+/// this is harder than it looks like at first.
+///
+/// As panicking from within a signal handler would be a panic across FFI boundary (which is
+/// undefined behavior), the passed handler must not panic.
+///
+/// If you find these limitations hard to satisfy, choose from the helper functions in submodules
+/// of this library ‒ these provide safe interface to use some common signal handling patters.
+///
+/// # Race condition
+///
+/// Currently, there's a short race condition. If this is the first action for the given signal,
+/// there was another signal handler previously and the signal comes into a different thread during
+/// this function, it can happen the original handler is not chained in this one instance.
+///
+/// This is considered unimportant problem, since most programs install their signal handlers
+/// during startup, before the signals effectively do anything. If you want to avoid the race
+/// condition completely, initialize all signal handling before starting any threads.
+///
+/// # Performance
+///
+/// Even when it is possible to repeatedly install and remove actions during the lifetime of a
+/// program, the installation and removal is considered a slow operation and should not be done
+/// very often. Also, there's limited (though huge) amount of distinct IDs (they are `u64`).
+///
+/// # Examples
+///
+/// ```rust
+/// extern crate signal_hook;
+///
+/// use std::io::Error;
+/// use std::process;
+///
+/// fn main() -> Result<(), Error> {
+///     let signal = unsafe { signal_hook::register(signal_hook::SIGTERM, || process::abort()) }?;
+///     // Stuff here...
+///     signal_hook::unregister(signal); // Not really necessary.
+///     Ok(())
+/// }
+/// ```
+pub unsafe fn register<F>(signal: c_int, action: F) -> Result<SigId, Error>
+where
+    F: Fn() + Sync + Send + 'static,
+{
+    assert!(
+        !FORBIDDEN.contains(&signal),
+        "Attempted to register forbidden signal {}",
+        signal,
+    );
+    let data = GlobalData::ensure();
+    let mut state = data.state.lock().unwrap();
+    let action_id = ActionId(state.action_id);
+    state.action_id += 1;
+
+    let sig_id = SigId {
+        signal,
+        action: action_id,
+    };
+
+    let all_signals = state
+        .all_signals
+        .entry(signal)
+        .or_insert_with(BTreeMap::new);
+    all_signals.insert(action_id, Box::new(action));
+
+    Ok(sig_id)
+}
```

Comment:
Reading this function, I have the impression there must be something missing. Who registers the `handler` function into the OS?

Reply:
Unlike `signal`/`sigaction`, `SetConsoleCtrlHandler` registry is not signal-specific. Therefore I registered the handler at `GlobalData` initialization.